salt/modules/win_event_viewer.py

```diff
@@ -0,0 +1,285 @@
+# -*- coding: utf-8 -*-
+'''
+:codeauthor: Charles McMarrow <cmcmarrow@saltstack.com>
+gives access to Windows event log
+'''
+
+# Import Third Party Libs
+import win32evtlog
+import win32evtlogutil
+import winerror
+import pywintypes
+
+# keys of all the parts of a Event supported by the API
+EVENT_PARTS = ('closingRecordNumber',
+               'computerName',
+               'data',
+               'eventCategory',
+               'eventID',
+               'eventType',
+               'recordNumber',
+               'reserved',
+               'reservedFlags',
+               'sid',
+               'sourceName',
+               'stringInserts',
+               'timeGenerated',
+               'timeWritten')
+
+# key to position in time tuple
+TIME_PARTS = {"year": 0,
+              "month": 1,
+              "day": 2,
+              "hour": 3,
+              "minute": 4,
+              "second": 5}
+
+
+def _change_str_to_bytes(data, encoding='utf-8', encode_keys=False):
+    '''
+    info: will make string objects into byte objects
+    warning: this function will destroy the data object and objects that data links to
+    :param data: object
+    :param encoding: str
+    :param encode_keys: bool: if false key strings will not be turned into bytes
+    :return: new object
+    '''
+
+    if isinstance(data, dict):
+        new_dict = {}
+        # recursively check every item in dict
+        for key in data:
+            item = _change_str_to_bytes(data.get(key), encoding)
+            if encode_keys:
+                # keys that are strings most be made into bytes
+                key = _change_str_to_bytes(key, encoding)
+            new_dict[key] = item
+        data = new_dict
+    elif isinstance(data, list):
+        new_list = []
+        # recursively check every item in list
+        for item in data:
+            new_list.append(_change_str_to_bytes(item, encoding))
+        data = new_list
+    elif isinstance(data, tuple):
+        new_list = []
+        # recursively check every item in list
+        for item in data:
+            new_list.append(_change_str_to_bytes(item, encoding))
+        data = tuple(new_list)
+    elif isinstance(data, str):
+        # data is turning into bytes because if was a string
+        data = data.encode(encoding)
+
+    return data
+
+
+def _get_raw_time(time):
+    '''
+    info: will make a pywintypes.datetime into a tuple
+    :param time: pywintypes.datetime
+    :return: tuple
+    '''
```

Comment:
I can. I'm just trying to cut down on memory usage. Some even logs can get very large and adding another object can add up relay quickly.

Reply:
Some people will be working with 100,000 events

Reply:
That's a fair point. I did give this a shot (in Python3):

```
import tracemalloc
from datetime import datetime

tracemalloc.start()
snapshot1 = tracemalloc.take_snapshot()
dates2 = [datetime(1,2,3,4,5,6,7) for _ in range(100_000)]
dates = [(1,2,3,4,5,6,7) for _ in range(100_000)]
snapshot2 = tracemalloc.take_snapshot()

top_stats = snapshot2.compare_to(snapshot1, 'lineno')
for stat in top_stats[:10]:
    print(stat)
```

It *is* quite a bit more mem  (800k -> 4000k), for 100,000. It might be worth it. What do you think? Though at 100,000,000 it's unlikely to be worth it :joy: It's already taken quite a bit of time to try that....

finished with a nice linear increase. 

One potential compromise might be to do a `namedtuple`. At least then we could still access with year/month/day/hour/minute/second properties. From my experiment it doesn't seem to have changed much:

```
/private/tmp/what.py:11: size=806 KiB (+806 KiB), count=3 (+3), average=269 KiB
/private/tmp/what.py:10: size=806 KiB (+806 KiB), count=2 (+2), average=403 KiB```

Reply:
also every event has two pywintypes.datetime that can be different

Reply:
I'll look at namedtuple. I agree it might be a good compromise.

Reply:
I think 100,000,000 it's unlikely too because of default caps.